presto-spi/src/main/java/com/facebook/presto/spi/PageBuilder.java

```diff
@@ -77,14 +78,20 @@ private PageBuilder(int initialExpectedEntries, int maxPageBytes, List<? extends
             }
         }
 
-        blockBuilders = new BlockBuilder[types.size()];
-        for (int i = 0; i < blockBuilders.length; i++) {
-            blockBuilders[i] = types.get(i).createBlockBuilder(
-                    pageBuilderStatus.createBlockBuilderStatus(),
-                    expectedEntries,
-                    pageBuilderStatus.getMaxBlockSizeInBytes() / expectedEntries);
-        }
         this.initialExpectedEntries = expectedEntries;
+
+        if (builders.isPresent()) {
+            blockBuilders = builders.get();
```

Comment:
Instead of creating the new `BlockBuilder`s in `resetNew`, I think we should (maybe *must*) create them here.  In that case, we should rename builders, to `templateBlockBuilders`.